File: scripts/upload-to-pinecone.ts
import { Pinecone } from '@pinecone-database/pinecone';
import { CohereClient } from 'cohere-ai';
import { readFile } from 'fs/promises';
import path from 'path';
import * as dotenv from 'dotenv';

// Load environment variables
dotenv.config({ path: '.env.local' });

const pinecone = new Pinecone({
  apiKey: process.env.PINECONE_API_KEY!,
});

const cohere = new CohereClient({
  token: process.env.COHERE_API_KEY!,
});

interface Chunk {
  id: string;
  text: string;
  metadata: {
    centre: string;
    category: string;
    chunk_index: number;
    total_chunks: number;
  };
}

// Smart chunking function
function smartChunk(text: string, maxChunkSize: number): string[] {
  const sentences = text.split(/[.!?]+/).filter(s => s.trim());
  const chunks: string[] = [];
  let currentChunk = '';

  for (const sentence of sentences) {
    const trimmedSentence = sentence.trim();
    if (!trimmedSentence) continue;

    if ((currentChunk + trimmedSentence).length > maxChunkSize && currentChunk) {
      chunks.push(currentChunk.trim());
      currentChunk = trimmedSentence;
    } else {
      currentChunk += (currentChunk ? '. ' : '') + trimmedSentence;
    }
  }

  if (currentChunk.trim()) {
    chunks.push(currentChunk.trim());
  }

  return chunks.length > 0 ? chunks : [text];
}

// Extract centre name from ## heading
function extractCentreName(line: string): string | null {
  const match = line.match(/^##\s+(.+)$/);
  return match ? match[1].trim() : null;
}

// Extract category from ### heading
function extractCategory(line: string): string | null {
  const match = line.match(/^###\s+(.+)$/);
  return match ? match[1].trim() : null;
}

// Normalize text for ID generation
function normalizeForId(text: string): string {
  return text
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '');
}

async function processKnowledgeBase() {
  console.log('🔍 Reading knowledge base...');
  const filePath = path.join(process.cwd(), 'data', 'knowledge-base-clean.md');
  const content = await readFile(filePath, 'utf-8');

  console.log(`📄 File size: ${content.length} characters`);

  const lines = content.split('\n');
  console.log(`📄 Total lines: ${lines.length}`);
  const chunks: Chunk[] = [];

  let currentCentre = '';
  let currentCategory = '';
  let currentContent = '';

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i].trim();

    // Check for centre heading (##)
    const centreName = extractCentreName(line);
    if (centreName) {
      console.log(`🏢 Found centre: ${centreName}`);
      // Process any accumulated content before changing centre
      if (currentContent.trim() && currentCentre && currentCategory) {
        processContent(currentCentre, currentCategory, currentContent, chunks);
        currentContent = '';
      }
      currentCentre = centreName;
      currentCategory = '';
      continue;
    }

    // Check for category heading (###)
    const categoryName = extractCategory(line);
    if (categoryName) {
      console.log(`  📁 Found category: ${categoryName}`);
      // Process any accumulated content before changing category
      if (currentContent.trim() && currentCentre && currentCategory) {
        processContent(currentCentre, currentCategory, currentContent, chunks);
        currentContent = '';
      }
      currentCategory = categoryName;
      continue;
    }

    // Accumulate content
    if (currentCentre && currentCategory) {
      currentContent += line + ' ';
    }
  }

  // Process final content
  if (currentContent.trim() && currentCentre && currentCategory) {
    console.log(`✅ Processing final content for ${currentCentre} - ${currentCategory}`);
    processContent(currentCentre, currentCategory, currentContent, chunks);
  }

  console.log(`\n📊 Total chunks created: ${chunks.length}`);
  console.log(`\n🔄 Generating embeddings and uploading to Pinecone...`);

  // Get the Pinecone index
  const index = pinecone.index(process.env.PINECONE_INDEX_NAME!);

  // Process in batches (Cohere allows up to 96 texts per embed call)
  const batchSize = 96;
  for (let i = 0; i < chunks.length; i += batchSize) {
    const batch = chunks.slice(i, i + batchSize);
    console.log(`\n📤 Processing batch ${Math.floor(i / batchSize) + 1}/${Math.ceil(chunks.length / batchSize)}`);
    console.log(`   Chunks ${i + 1} to ${Math.min(i + batchSize, chunks.length)} of ${chunks.length}`);

    // Generate embeddings for the batch
    const texts = batch.map(chunk => chunk.text);
    const embeddingResponse = await cohere.embed({
      texts,
      model: 'embed-english-v3.0',
      inputType: 'search_document',
    });

    // Prepare vectors for Pinecone
    const vectors = batch.map((chunk, index) => ({
      id: chunk.id,
      values: embeddingResponse.embeddings[index],
      metadata: {
        ...chunk.metadata,
        text: chunk.text,
      },
    }));

    // Upsert to Pinecone
    await index.upsert(vectors);
    console.log(`   ✅ Uploaded ${vectors.length} vectors`);

    // Small delay to avoid rate limits
    await new Promise(resolve => setTimeout(resolve, 100));
  }

  console.log(`\n✅ Successfully uploaded ${chunks.length} chunks to Pinecone!`);
}

function processContent(
  centre: string,
  category: string,
  content: string,
  chunks: Chunk[]
): void {
  const trimmedContent = content.trim();
  if (!trimmedContent) {
    console.log(`    ⚠️  Empty content for ${centre} - ${category}`);
    return;
  }

  // Create smaller chunks for better retrieval
  const textChunks = smartChunk(trimmedContent, 400);
  console.log(`    ✨ Creating ${textChunks.length} chunks for ${centre} - ${category}`);

  textChunks.forEach((chunkText, index) => {
    const centreId = normalizeForId(centre);
    const categoryId = normalizeForId(category);
    const chunkId = `${centreId}-${categoryId}-chunk-${index}`;

    // Add context to chunk text for better understanding
    const enhancedText = `${centre} ${category}: ${chunkText}`;

    chunks.push({
      id: chunkId,
      text: enhancedText,
      metadata: {
        centre: centreId,
        category: categoryId,
        chunk_index: index,
        total_chunks: textChunks.length,
      },
    });
  });
}

processKnowledgeBase().catch(console.error);

File: scripts/upload-to-pinecone.py
import os
import json
from pathlib import Path
from pinecone import Pinecone, ServerlessSpec
import cohere
import time
from dotenv import load_dotenv

# Load environment variables from .env.local
load_dotenv('.env.local')

# Initialize clients
pinecone = Pinecone(api_key=os.environ.get('PINECONE_API_KEY'))
cohere_client = cohere.ClientV2(api_key=os.environ.get('COHERE_API_KEY'))

# Configuration
INDEX_NAME = 'centre2'
DIMENSION = 1024  # Cohere embed-english-v3.0 dimension
BATCH_SIZE = 25  # Process in batches (rate limit is 100k tokens/min)

def chunk_text(text, max_chunk_size=1000, overlap=200):
    """Split text into overlapping chunks"""
    chunks = []
    lines = text.split('\n')

    current_chunk = []
    current_size = 0

    for line in lines:
        line_size = len(line)

        if current_size + line_size > max_chunk_size and current_chunk:
            # Save current chunk
            chunks.append('\n'.join(current_chunk))

            # Start new chunk with overlap
            overlap_lines = []
            overlap_size = 0
            for prev_line in reversed(current_chunk):
                if overlap_size + len(prev_line) <= overlap:
                    overlap_lines.insert(0, prev_line)
                    overlap_size += len(prev_line)
                else:
                    break

            current_chunk = overlap_lines
            current_size = overlap_size

        current_chunk.append(line)
        current_size += line_size

    # Add the last chunk
    if current_chunk:
        chunks.append('\n'.join(current_chunk))

    return chunks

def extract_category_from_chunk(chunk_text):
    """Try to extract category from chunk headers"""
    lines = chunk_text.split('\n')
    for line in lines[:5]:  # Check first 5 lines
        if line.startswith('##'):
            category = line.replace('#', '').strip()
            # Simplify category names
            if 'official' in category.lower() or 'pricing' in category.lower():
                return 'pricing-schedules'
            elif 'website' in category.lower() or 'facilities' in category.lower():
                return 'facilities-programs'
            elif 'opening' in category.lower() or 'hours' in category.lower():
                return 'hours'
            elif 'contact' in category.lower():
                return 'contact'
            else:
                return category.lower()[:50]  # Limit length
    return 'general'

def process_centre_file(centre_id, file_path, metadata_path):
    """Process a single centre's combined file into chunks"""
    print(f"Processing {centre_id}...")

    # Read the combined markdown
    with open(file_path, 'r', encoding='utf-8') as f:
        content = f.read()

    # Read metadata for URLs
    metadata = {'pages': []}
    if metadata_path.exists():
        try:
            with open(metadata_path, 'r', encoding='utf-8') as f:
                metadata = json.load(f)
        except:
            pass

    # Chunk the content
    chunks = chunk_text(content)
    print(f"  Created {len(chunks)} chunks")

    # Prepare vectors with metadata
    vectors = []
    for i, chunk in enumerate(chunks):
        category = extract_category_from_chunk(chunk)

        vector_data = {
            'id': f"{centre_id}-{i}",
            'text': chunk,
            'metadata': {
                'centre': centre_id,
                'category': category,
                'chunk_index': i,
                'text': chunk[:1000]  # Store first 1000 chars in metadata
            }
        }
        vectors.append(vector_data)

    return vectors

def upload_to_pinecone(vectors_batch):
    """Upload a batch of vectors to Pinecone"""
    # Generate embeddings for all texts
    texts = [v['text'] for v in vectors_batch]

    print(f"  Generating embeddings for {len(texts)} chunks...")
    response = cohere_client.embed(
        texts=texts,
        model='embed-english-v3.0',
        input_type='search_document'
    )

    embeddings = response.embeddings.float_

    # Prepare upsert data
    upsert_data = []
    for i, vector_data in enumerate(vectors_batch):
        upsert_data.append({
            'id': vector_data['id'],
            'values': embeddings[i],
            'metadata': vector_data['metadata']
        })

    # Upsert to Pinecone
    index_host = 'centre2-te9sgjq.svc.aped-4627-b74a.pinecone.io'
    index = pinecone.Index(INDEX_NAME, host=index_host)

    print(f"  Uploading {len(upsert_data)} vectors to Pinecone...")
    index.upsert(vectors=upsert_data)

    return len(upsert_data)

def main():
    combined_dir = Path('c:/Projects/centre2/data/combined')
    metadata_dir = Path('c:/Projects/centre2/data/crawled-deep')

    print("="*60)
    print("Uploading Combined Knowledge Base to Pinecone")
    print("="*60)
    print(f"Source: {combined_dir}")
    print(f"Index: {INDEX_NAME}")
    print(f"Embedding model: Cohere embed-english-v3.0")
    print()

    # Get all combined markdown files
    md_files = sorted(combined_dir.glob('*.md'))
    print(f"Found {len(md_files)} centre files\n")

    all_vectors = []
    total_chunks = 0

    # Process each centre
    for md_file in md_files:
        centre_id = md_file.stem
        metadata_path = metadata_dir / f"{centre_id}_metadata.json"

        vectors = process_centre_file(centre_id, md_file, metadata_path)
        all_vectors.extend(vectors)
        total_chunks += len(vectors)

    print(f"\nTotal chunks to upload: {total_chunks}")
    print(f"Processing in batches of {BATCH_SIZE}...")

    # Upload in batches
    uploaded_count = 0
    for i in range(0, len(all_vectors), BATCH_SIZE):
        batch = all_vectors[i:i + BATCH_SIZE]
        batch_num = i // BATCH_SIZE + 1
        total_batches = (len(all_vectors) + BATCH_SIZE - 1) // BATCH_SIZE

        print(f"\nBatch {batch_num}/{total_batches}:")
        count = upload_to_pinecone(batch)
        uploaded_count += count

        # Rate limiting - wait between batches to stay under 100k tokens/min
        if i + BATCH_SIZE < len(all_vectors):
            print("  Waiting 10 seconds...")
            time.sleep(10)

    print("\n" + "="*60)
    print(f"Upload Complete!")
    print(f"  Total vectors uploaded: {uploaded_count}")
    print(f"  Centres processed: {len(md_files)}")
    print("="*60)

if __name__ == '__main__':
    main()
